File: src/matrix.rs
use crate::m::BigM;
use std::cmp;

#[derive(Debug, Eq, Clone)]
pub struct Matrix {
    pub rows: usize,
    pub columns: usize,
    values: Vec<Vec<BigM>>,
}

impl Matrix {
    pub fn new(rows: usize, columns: usize) -> Matrix {
        Matrix {rows, columns, values: Vec::<Vec::<BigM>>::with_capacity(rows)}
    }

    pub fn new_with_data(rows: usize, columns: usize, values: Vec<Vec<BigM>>) -> Matrix {
        Matrix { rows, columns, values}
    }

    pub fn get(&self, row: usize, column: usize) -> BigM {
        self.values[row][column]
    }

    pub fn set(&mut self, row: usize, column: usize, value: BigM) -> bool {
        if row > self.rows-1 { 
            return false;
        } else if column > self.columns-1 {
            return false;
        }
        self.values[row][column] = value;
        return true;
    }
}

impl cmp::PartialEq<Matrix> for Matrix {
    fn eq(&self, other: &Matrix) -> bool {
        return self.values == other.values;
    }

    fn ne(&self, other: &Matrix) -> bool {
        return !(self.values == other.values);
    }
}

File: src/lib/matrix.c
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include "../../include/bigm.h"


typedef struct Matrix {
    int rows;
    int columns;
    BigM **values;
} Matrix;

void free_matrix(Matrix *mat) {
    for (int i=0; i < mat->rows; i++) {
        free(mat->values[i]);
    }
    free(mat->values);

    free(mat);
}

Matrix* new_matrix(int rows, int columns) {
    Matrix* mat = malloc(sizeof(Matrix));
    if (mat == NULL) {
        return NULL;
    }
    mat->rows = rows;
    mat->columns = columns;
    mat->values = malloc(sizeof(BigM**) * rows);
    if (mat->values == NULL) {
        free(mat);
        return NULL;
    }
    for (int i = 0; i < rows; i++) {
        mat->values[i] = malloc(sizeof(BigM) * columns);
        if (mat->values[i] == NULL) {
            for (int j = 0; j < i; j++) {
                free(mat->values[j]);
            }
            free(mat->values);
            free(mat);
            return NULL;
        }
    }
    return mat;
}

Matrix* new_matrix_with_data(int rows, int columns, BigM** data) {
    Matrix* mat = new_matrix(rows, columns);
    if (mat == NULL) {
        return NULL;
    }
    for (int i = 0; i < rows; i++) {
        memcpy(mat->values[i], data[i], sizeof(BigM) * columns);
    }
    return mat;
}

BigM get(const Matrix* mat, int row, int column) {
    return mat->values[row][column];
}

bool set(Matrix* mat, int row, int column, BigM value) {
    if (row > mat->rows-1) { 
        return false;
    } else if (column > mat->columns-1) {
        return false;
    }
    mat->values[row][column] = value;
    return true;
}

void print_mat(Matrix *mat) {
    char *str = malloc(30 * sizeof(char));
    for (int i=0; i < mat->rows; i++) {
        for (int j=0; j < mat->columns; j++) {
            BigM_str(get(mat, i, j), str);
            printf("%s ", str);
        }
        printf("\n");
    }
    free(str);
}